syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";
package loyalty;


service PromoCode {
  rpc CreateV1 (CreatePromoCodeRequestV1) returns (CreatePromoCodeResponseV1) {}
  rpc ReserveV1 (CommonPromoCodeRequestV1) returns (CommonResponseV1) {}
  rpc FreeV1 (ReserveIdRequestV1) returns (google.protobuf.Empty) {}
  rpc ApplyV1 (ReserveIdRequestV1) returns (google.protobuf.Empty) {}
}

service Discount {
  rpc CreateV1 (CreateDiscountRequestV1) returns (CommonResponseV1) {}
  rpc ApplyV1 (ApplyDiscountRequestV1) returns (ApplyDiscountResponseV1) {}
}

message CreatePromoCodeRequestV1 {
  string code = 1;
  int32 discount_percents = 2;
  google.protobuf.Timestamp expired_at = 3;
  repeated int64 users_ids = 4;
  bool send_email = 5;
}

message CreatePromoCodeResponseV1 {
  string id = 1;
  string code = 2;
  int32 discount_percents = 3;
  google.protobuf.Timestamp expired_at = 4;
  repeated int64 users_ids = 5;
}

message CommonResponseV1 {
  string id = 1;
}

message CommonPromoCodeRequestV1 {
  string code = 1;
  int64 user_id = 2;
}

message ReserveIdRequestV1 {
  string reserve_id = 1;
}

message CreateDiscountRequestV1 {
  int64 user_id = 1;
  int32 discount_percents = 2;
   google.protobuf.Timestamp expired_at = 3;
}

message ApplyDiscountRequestV1 {
  int64 user_id = 1;
}

message ApplyDiscountResponseV1 {
  int32 discount_percents = 1;
}